src/HttpClient/HttpClientFactory.php

```diff
@@ -0,0 +1,116 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Sentry\HttpClient;
+
+use Http\Client\Common\Plugin\AuthenticationPlugin;
+use Http\Client\Common\Plugin\BaseUriPlugin;
+use Http\Client\Common\Plugin\DecoderPlugin;
+use Http\Client\Common\Plugin\ErrorPlugin;
+use Http\Client\Common\Plugin\HeaderSetPlugin;
+use Http\Client\Common\Plugin\RetryPlugin;
+use Http\Client\Common\PluginClient;
+use Http\Client\Curl\Client as HttpCurlClient;
+use Http\Client\HttpAsyncClient as HttpAsyncClientInterface;
+use Http\Discovery\HttpAsyncClientDiscovery;
+use Http\Message\StreamFactory;
+use Http\Message\UriFactory as UriFactoryInterface;
+use Sentry\HttpClient\Authentication\SentryAuthentication;
+use Sentry\HttpClient\Plugin\GzipEncoderPlugin;
+use Sentry\Options;
+
+/**
+ * Default implementation of the {@HttpClientFactoryInterface} interface that uses
+ * Httplug to autodiscover the HTTP client if none is passed by the user.
+ */
+final class HttpClientFactory implements HttpClientFactoryInterface
+{
+    /**
+     * @var UriFactoryInterface The PSR-7 URI factory
+     */
+    private $uriFactory;
+
+    /**
+     * @var StreamFactory The PSR-7 stream factory
+     */
+    private $streamFactory;
+
+    /**
+     * @var HttpAsyncClientInterface|null The HTTP client
+     */
+    private $httpClient;
+
+    /**
+     * @var string The name of the SDK
+     */
+    private $sdkIdentifier;
+
+    /**
+     * @var string The version of the SDK
+     */
+    private $sdkVersion;
+
+    /**
+     * Constructor.
+     *
+     * @param UriFactoryInterface           $uriFactory    The PSR-7 URI factory
+     * @param StreamFactory                 $streamFactory The PSR-17 stream factory
+     * @param HttpAsyncClientInterface|null $httpClient    The HTTP client
+     * @param string                        $sdkIdentifier The SDK identifier
+     * @param string                        $sdkVersion    The SDK version
+     */
+    public function __construct(UriFactoryInterface $uriFactory, StreamFactory $streamFactory, ?HttpAsyncClientInterface $httpClient, string $sdkIdentifier, string $sdkVersion)
+    {
+        $this->uriFactory = $uriFactory;
+        $this->httpClient = $httpClient;
+        $this->sdkIdentifier = $sdkIdentifier;
+        $this->sdkVersion = $sdkVersion;
+        $this->streamFactory = $streamFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function create(Options $options): HttpAsyncClientInterface
+    {
+        if (null === $options->getDsn()) {
+            throw new \RuntimeException('Cannot create an HTTP client without the Sentry DSN set in the options.');
+        }
+
+        $httpClient = $this->httpClient;
+
+        if (null !== $httpClient && null !== $options->getHttpProxy()) {
```

Comment:
I would check if the client is a compatible one, instead of `null` altogether.

Reply:
We could, but the result would not be the same. The whole logic is:

- If a HTTP client is set and the `http_proxy` option is set then throw an error because we cannot modify the options of an already instantiated client at runtime
- If a HTTP client is not set and the `http_proxy` option is set then try to instantiate a Guzzle client or a Httplug Curl client in this order
- If a HTTP client is set and no `http_proxy` option is set then use it
- In all other cases try to use autodiscovery or throw if we cannot find a suitable HTTP client

As you can see the logic cannot be replaced with a simple `if` that compares whether the HTTP client is supported out-of-the-box or not as the conditions to check are multiple